File: mobile/android/FUN/app/src/main/java/com/fun/app/data/models/Transaction.kt
/**
 * FUN App - Transaction Models
 */

package com.fun.app.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Transaction(
    @Json(name = "_id") val id: String,
    val type: String,  // "credit_purchase", "credit_spend", "subscription"
    val amount: Double,
    val currency: String,
    val paymentMethod: String,  // "google_play", "stripe", "system"
    val status: String,
    val metadata: TransactionMetadata?,
    val createdAt: String
)

@JsonClass(generateAdapter = true)
data class TransactionMetadata(
    val credits: Int?,
    val seriesId: String?,
    val episodeNum: Int?,
    val productId: String?
)

@JsonClass(generateAdapter = true)
data class TransactionsResponse(
    val transactions: List<Transaction>,
    val pagination: Pagination
)

@JsonClass(generateAdapter = true)
data class CreditProduct(
    @Json(name = "_id") val id: String,
    val name: String,
    val credits: Int,
    val priceUSD: Double,
    val stripePriceId: String?,
    val appleProductId: String?,
    val googleProductId: String?,
    val bonus: Int?
) {
    val isPopular: Boolean
        get() = credits == 500

    val hasBonus: Boolean
        get() = bonus != null && bonus > 0
}

@JsonClass(generateAdapter = true)
data class CreditProductsResponse(
    val products: List<CreditProduct>
)

@JsonClass(generateAdapter = true)
data class SubscriptionProduct(
    @Json(name = "_id") val id: String,
    val name: String,
    val tier: String,  // "monthly", "annual"
    val priceUSD: Double,
    val stripePriceId: String?,
    val appleProductId: String?,
    val googleProductId: String?,
    val features: List<String>
)

@JsonClass(generateAdapter = true)
data class SubscriptionProductsResponse(
    val products: List<SubscriptionProduct>
)

@JsonClass(generateAdapter = true)
data class VerifyGoogleIAPRequest(
    val purchaseToken: String,
    val productId: String,
    val episodeId: String? = null
)

File: mobile/ios/FUN/FUN/Models/Transaction.swift
/**
 * FUN App - Transaction Model
 */

import Foundation

struct Transaction: Codable, Identifiable {
    let id: String
    let type: String  // "credit_purchase", "credit_spend", "subscription"
    let amount: Double
    let currency: String
    let paymentMethod: String  // "apple_iap", "stripe", "system"
    let status: String
    let metadata: TransactionMetadata?
    let createdAt: String
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case type, amount, currency, paymentMethod, status, metadata, createdAt
    }
}

struct TransactionMetadata: Codable {
    let credits: Int?
    let seriesId: String?
    let episodeNum: Int?
    let productId: String?
}

struct TransactionsResponse: Codable {
    let transactions: [Transaction]
    let pagination: Pagination
}

struct CreditProduct: Codable, Identifiable {
    let id: String
    let name: String
    let credits: Int
    let priceUSD: Double
    let stripePriceId: String?
    let appleProductId: String?
    let bonus: Int?
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, credits, priceUSD, stripePriceId, appleProductId, bonus
    }
    
    var isPopular: Bool {
        return credits == 500
    }
    
    var hasBonus: Bool {
        return bonus != nil && bonus! > 0
    }
}

struct CreditProductsResponse: Codable {
    let products: [CreditProduct]
}

struct SubscriptionProduct: Codable, Identifiable {
    let id: String
    let name: String
    let tier: String  // "monthly", "annual"
    let priceUSD: Double
    let stripePriceId: String?
    let appleProductId: String?
    let features: [String]
    
    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, tier, priceUSD, stripePriceId, appleProductId, features
    }
}

struct SubscriptionProductsResponse: Codable {
    let products: [SubscriptionProduct]
}
